zebra-state/src/service/finalized_state.rs

```diff
@@ -329,10 +337,15 @@ impl FinalizedState {
                 };
                 batch.zs_insert(tx_by_hash, transaction_hash, transaction_location);
 
-                // Mark all transparent inputs as spent
+                // Mark all transparent inputs as spent, collect them as well.
                 for input in transaction.inputs() {
                     match input {
                         transparent::Input::PrevOut { outpoint, .. } => {
+                            all_utxos_spent_by_block.insert(
+                                *outpoint,
+                                self.utxo(outpoint)
+                                    .expect("Utxo for OutPoint should be present"),
```

Comment:
I think `service::check::tests::utxo::accept_later_transparent_spend_from_this_block` is failing because a block can:
* create a UTXO, then
* spend that transparent output in a later transaction in the block.

This happens because UTXOs created in this block are inserted as part of a delayed batch, which actually runs here:
https://github.com/ZcashFoundation/zebra/blob/76591ceeede5d917ad6db54ae4fce12d67cddd2d/zebra-state/src/service/finalized_state.rs#L397

But `self.utxo(outpoint)` does the lookup straight away (before the block batch is committed).

`new_outputs` should have the missing UTXOs from this block.

Reply:
We opened ticket #2588 to fix this confusing batch behaviour.
I've updated that ticket to mention UTXOs as well.

(I think it's confused everyone who has worked on this part of the state.)

Reply:
Ok, this makes sense, thank you. I joined the two hashmaps at https://github.com/ZcashFoundation/zebra/pull/2599/commits/ad517b85d3286a2f065e93d09bf9a56ba9bd992f

However it is not fixing the test cases, i will try to find out more.

Reply:
Ok, so, in the inputs loop there was a panic i added for when the outpoint is not found whoever this seems to be not fatal.

By removing that panic all the tests pass.

https://github.com/ZcashFoundation/zebra/pull/2599/commits/bfcc1949681348aa0538c0ed1550c0ecfe9df414